### PR TITLE
[Validator] Clarify UUID validator behavior

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24209
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

TLDR; after some investigation this is now a doc update.

This fixes the variant detection of a UUID, in strict validation, as described by https://tools.ietf.org/html/rfc4122#section-4.1.1

Tested variant UUID's taken from [ramsey/uuid](https://github.com/ramsey/uuid/blob/master/tests/UuidTest.php#L506). I cannot produce an invalid variant UUID anymore, as i believe by now anything would be variant 3 (future reserved).

/cc @backbone87

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/UuidValidator.php [hunk 1]
```diff
@@ -308,35 +308,34 @@ private function validateStrict($value, Uuid $constraint)
...
             }
         }
 
-        // Check variant - first two bits must equal "10"
-        //   0b10xx
-        // & 0b1100 (12)
-        // = 0b1000 (8)
-        if (8 !== (hexdec($value[self::STRICT_VARIANT_POSITION]) & 12)) {
+        // Check version - applies to variant 1 and 2
+        if ((8 === ($variant & 12) || 12 === ($variant & 14)) && !in_array($value[self::STRICT_VERSION_POSITION], $constraint->versions)) {
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/UuidValidator.php]: https://en.wikipedia.org/wiki/Universally_unique_identifier#Versions

**Comment by nicolas-grekas** [comment]
Should we really allow variants 3? Being "future reserved" means they should not be used today, isn't it?

**Comment by ro0NL** [comment]
From the RFC pov it's a valid UUID, same as special nil variant.

I've updated the approach do to be more BC friendly, if you pass versions (default behavior) we imply a variant 1 or 2. Thus INVALID_VARIANT_ERROR for variant 3 & 4 in this case, see tests.

In 3.x we could add a new variant constraint, with a default set to variant 1 & 2.

**Comment by backbone87** [comment]
puhhh, when i tested it, it reported my UUIDs from the AD still as invalid (in strict mode).
then i noticed the wikipedia stuff about byte order of MS GUIDs when transmitted in binary form and as it happens the LDAP entry of the AD contains the objectGUIDs in binary form.

After switching to the right codec of ramsey/uuid to decode/encode the GUIDs, they are being validated in strict mode. The thing is that they are valid with the old validator too, because they are now valid RFC variant UUIDs and not MS legacy UUIDs.

Just the byte order was f***ed up.

(Btw. the UUID i posted in the issue, was invalid because the byte order wasnt fixed, when decoding from binary).

So i cant really test this one, since my AD GUIDs are strictly valid with and without this PR

**Comment by ro0NL** [comment]
Thanks. First im not 100% into UUID's on its lowest level. This PR is a result after some reading and checking other implems. So basically i reverse engineered it from tests + php-src code; https://3v4l.org/DB4i9

Testing your input `d7bb0bc1-0b24-944c-bea2-ef7da43e5c5c` _still_ fails, it produces variant 1 with an invalid version. This is indeed about codecs, using the ramsey GuidStringCodec makes it a version 4.

```php
$u1 = \Ramsey\Uuid\Uuid::fromString('d7bb0bc1-0b24-944c-bea2-ef7da43e5c5c');
$u2 = (new \Ramsey\Uuid\UuidFactory(new \Ramsey\Uuid\FeatureSet($guid = true)))->fromString('d7bb0bc1-0b24-944c-bea2-ef7da43e5c5c');

dump(
    $u1->getVariant() === \Ramsey\Uuid\Uuid::RFC_4122,
    in_array($u1->getVersion(), array(1,2,3,4,5)), // version=9
    $u2->getVariant() === \Ramsey\Uuid\Uuid::RFC_4122,
    $u2->getVersion() === \Ramsey\Uuid\Uuid::UUID_TYPE_RANDOM // version=4
);
```

```
true
false
true
true
```

Basically SF is stuck at version=9. IMHO we should stop here (im quite convinced of this patch right now) and switch to proper libs in 3.x for better compatibility and validation control on symfony side, as using `$guid=true` on ramsey side also invalidates a valid version 4 for example.

We might drop a note in 2.7 and focus on 3.x behavior/api. Yet im quite proud of the patch :angel:

**Comment by backbone87** [comment]
your test isnt right. `d7bb0bc1-0b24-944c-bea2-ef7da43e5c5c` is a wrongly decoded UUID.

this will fix the UUIDs string representation:
```php
$u1 = \Ramsey\Uuid\Uuid::fromString('d7bb0bc1-0b24-944c-bea2-ef7da43e5c5c');
$u2 = (new \Ramsey\Uuid\UuidFactory(new \Ramsey\Uuid\FeatureSet($guid = true)))->fromBytes($u1->toBytes());
```

**Comment by ro0NL** [comment]
Agree :+1: 

```diff
 /**
- * Validates whether the value is a valid UUID per RFC 4122.
+ * Validates whether the value is a valid UUID (also known as GUID).
+ *
+ * Strict validation will allow a variant 1 UUID as described per RFC 4122. Loose validation will allow any type of UUID.
+ *
+ * For better compatibility, both loose and strict, you should consider using specialized a UUID library instead.
+ * A well known composer package is "ramsey/uuid".
  *
  * @author Colin O'Dell <colinodell@gmail.com>
  * @author Bernhard Schussek <bschussek@gmail.com>
  *
  * @see http://tools.ietf.org/html/rfc4122
  * @see https://en.wikipedia.org/wiki/Universally_unique_identifier
+ * @see https://github.com/ramsey/uuid
  */
```

?

**Comment by nicolas-grekas** [comment]
> you should consider using a specialized UUID library like "ramsey/uuid" instead.

:+1: otherwise
ping @colinodell

**Comment by backbone87** [comment]
i would fix the variant "indexing". despite the RFC listing isnt numbered, we should count from 1 (see also https://stackoverflow.com/a/6953207 )

so:
1.    0     x     x    Reserved, NCS backward compatibility.
2.    1     0     x    The variant specified in this document.
3.    1     1     0    Reserved, Microsoft Corporation backward compatibility
4.    1     1     1    Reserved for future definition.

so the strict mode validates variant 2 UUIDs

**Comment by ro0NL** [comment]
https://en.wikipedia.org/wiki/Universally_unique_identifier#Variants explicitly refers to variant 0, whereas stackoverflow implies (it's an `<ol>` really).

**Comment by backbone87** [comment]
wikipedia references https://opensource.apple.com/source/CF/CF-299.35/Base.subproj/uuid.c.auto.html when talking about variants. idk how relevant this file is

**Comment by ro0NL** [comment]
RFC has the latest copyright year =/ From apple source `There are three "variants" of UUIDs that this code knows about`. RFC has 4.

**Comment by backbone87** [comment]
better dont mention variant indexes at all then

**Comment by ro0NL** [comment]
It takes away the confusion which this RFC creates, and exposes the variant SF actually follows.

```
1     0     x    The variant specified in this document.
```

Written by RFC describing (listing) variants :disappointed:

But i see your point; ramsey/uuid describes this as a simple named value (`RFC_4122|RESERVED_MICROSOFT|...`). But is more confusing IMHO. So in terms of correctness and this is doc related i'd stick with "variant 1".

**Comment by backbone87** [comment]
i think there is no confusion with the RFC, it just lists 4 variants, but dont assign them any index, just a description for every variant and the bit signature

**Comment by ro0NL** [comment]
Updated.

**Comment by backbone87** [comment]
good for me

**Review by nicolas-grekas — APPROVED** [review]

**Comment by colinodell** [comment]
@nicolas-grekas Looks good to me :+1:

**Comment by nicolas-grekas** [comment]
Thank you @ro0NL.
